crates/kubelet-derive/src/lib.rs

```diff
@@ -0,0 +1,64 @@
+//! A crate for deriving state machine traits in Kubelet. Right now this crate only consists of a
+//! derive macro for the `TransitionTo` trait. In addition to the `derive` attribute, this macro
+//! also requires the use of a custom attribute called `transition_to` that specifies the types that
+//! can be transitioned to. Not specifying this attribute will result in a compile time error. A
+//! simple example of this is below:
+//!
+//! ```rust,no_run
+//! use kubelet_derive::TransitionTo;
+//!
+//! pub struct VolumeMount;
+//! pub struct ImagePullBackoff;
+//!
+//! #[derive(Default, Debug, TransitionTo)]
+//! #[transition_to(VolumeMount, ImagePullBackoff)]
+//! pub struct ImagePull;
+//!```
+
+extern crate proc_macro;
+
+use crate::proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput, Error, Meta};
+
+const ATTRIBUTE_NAME: &str = "transition_to";
+
+#[proc_macro_derive(TransitionTo, attributes(transition_to))]
+pub fn derive_transition_to(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let name = &input.ident;
+    let generics = &input.generics;
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let mut token_stream = TokenStream::new();
+
+    // We need to check if we found at least one attribute. If it isn't set, we need to error
+    let mut found_attr = false;
```

Comment:
I'd consider splitting this into a find/filter and an expand step a la:

```
let transition_to_attrs = filter_transition_to_attrs(input);
if transition_to_attrs.is_empty() {
  some_helpful_error()
} else {
  impl_tokens_for(transition_to_attrs)
}
```

This makes the high-level flow more self-documenting, and enables you to pull specific logical functions into named functions instead of having the "which inputs are interesting" + "what do I do with the interesting ones" all bound up in one deeply nested block.